### PR TITLE
POC implementation spec draft

### PR DESCRIPTION
**Pull request by hiddeco**
This PR is a POC implementation of the spec draft to be able to further draft the spec based on technical findings. It provides full coverage of common Helm actions while also allowing extensive configuration options per action, but is currently limited in the ways actions are triggered (i.e. rolling back on a test failure is not possible yet, but it does perform a rollback on upgrade failures).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+![overview](docs/diagrams/helm-controller-overview.png)
+
+Features:
+
+* watches for `HelmRelease` objects
+* fetches artifacts produced by [source-controller](https://github.com/fluxcd/source-controller)
+  from `Source` objects 
+* watches `Source` objects for revision changes
+* performs Helm actions as configured in the object
```
Comment (stefanprodan) [inline README.md:26]:
```suggestion
* performs Helm actions as configured in the `HelmRelease`
```

File: config/manager/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            memory: 512Mi
+          requests:
+            cpu: 100m
+            memory: 64Mi
+        volumeMounts:
+          - name: temp
+            mountPath: /tmp
+      volumes:
+        - name: temp
+          emptyDir: {}
```
Comment (stefanprodan) [inline config/manager/deployment.yaml:52]: New line here

File: config/manager/kustomization.yaml [hunk 3]
```diff
@@ -1,2 +1,8 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
 resources:
-- manager.yaml
+  - deployment.yaml
+images:
+  - name: fluxcd/helm-controller
+    newName: fluxcd/helm-controller
+    newTag: v0.0.1-alpha.1
```
Comment (stefanprodan) [inline config/manager/kustomization.yaml:8]: New line here

File: controllers/helmchart_watcher.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+	return ctrl.Result{}, nil
+}
+
+func (r *HelmChartWatcher) SetupWithManager(mgr ctrl.Manager) error {
+	// Create a HelmRelease index based on the HelmChart name
+	err := mgr.GetFieldIndexer().IndexField(context.TODO(), &v2.HelmRelease{}, v2.SourceIndexKey,
+		func(rawObj runtime.Object) []string {
+			hr := rawObj.(*v2.HelmRelease)
+			if hr.Spec.SourceRef.Kind == "HelmChart" {
```
Comment (stefanprodan) [inline controllers/helmchart_watcher.go:92]: I think we should add the kinds name to the their API packages as constants, we use strings in many places now, kustomize/notification/watchers/etc.

File: controllers/helmrelease_controller.go [hunk 5]
```diff
@@ -18,36 +18,376 @@ package controllers
...
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
 
+	sourcev1 "github.com/fluxcd/source-controller/api/v1alpha1"
```
Comment (stefanprodan) [inline controllers/helmrelease_controller.go:32]: Can you move this in the fluxcd group below

File: controllers/helmrelease_controller.go [hunk 6]
```diff
@@ -18,36 +18,376 @@ package controllers
...
+		WithEventFilter(HelmReleaseGarbageCollectPredicate{Config: r.Config, Log: r.Log}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: opts.MaxConcurrentReconciles}).
 		Complete(r)
 }
+
+func (r *HelmReleaseReconciler) release(log logr.Logger, hr v2.HelmRelease, source sourcev1.Source) (v2.HelmRelease, error) {
+	// Acquire lock
+	unlock, err := lock(fmt.Sprintf("%s-%s", hr.GetName(), hr.GetNamespace()))
+	if err != nil {
+		err = fmt.Errorf("tmp dir error: %w", err)
```
Comment (stefanprodan) [inline controllers/helmrelease_controller.go:169]:
```suggestion
		err = fmt.Errorf("lockfile error: %w", err)
```
I need to change this in kustomize controller as well.

File: go.mod [hunk 7]
```diff
@@ -3,10 +3,16 @@ module github.com/fluxcd/helm-controller
 go 1.13
 
 require (
+	github.com/fluxcd/pkg v0.0.2
```
Comment (stefanprodan) [inline go.mod:6]:
```suggestion
	github.com/fluxcd/pkg v0.0.3
```

File: go.mod [hunk 8]
```diff
@@ -3,10 +3,16 @@ module github.com/fluxcd/helm-controller
...
+	github.com/fluxcd/source-controller v0.0.2
```
Comment (stefanprodan) [inline go.mod:7]:
```suggestion
	github.com/fluxcd/source-controller v0.0.3
```

File: controllers/helmrelease_controller.go (hunk 6 above)
Comment (hiddeco) [reply controllers/helmrelease_controller.go:169]: It is quite likely that this same wrong message exists in the kustomize-controller.

**Pull request merged**